server_ext_test.go

```diff
@@ -97,3 +103,90 @@ func (s) TestServer_MaxHandlers(t *testing.T) {
 		t.Fatal("Received unexpected RPC error:", err)
 	}
 }
+
+// Tests the case where the stream worker goroutine option is enabled, and a
+// number of RPCs are initiated around the same time that Stop() is called. This
+// used to result in a write to a closed channel. This test verifies that there
+// is no panic.
+func (s) TestStreamWorkers_RPCsAndStop(t *testing.T) {
+	ss := stubserver.StartTestService(t, nil, grpc.NumStreamWorkers(uint32(runtime.NumCPU())))
+	// This deferred stop takes care of stopping the server when one of the
+	// below grpc.Dials fail, and the test exits early.
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	const numChannels = 20
+	const numRPCLoops = 20
+
+	// Create a bunch of clientconns and ensure that they are READY by making an
+	// RPC on them.
+	ccs := make([]*grpc.ClientConn, numChannels)
+	for i := 0; i < numChannels; i++ {
+		var err error
+		ccs[i], err = grpc.Dial(ss.Address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			t.Fatalf("[iteration: %d] grpc.Dial(%s) failed: %v", i, ss.Address, err)
+		}
+		defer ccs[i].Close()
+		client := testgrpc.NewTestServiceClient(ccs[i])
+		if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true)); err != nil {
+			t.Fatalf("EmptyCall() failed: %v", err)
+		}
+	}
+
+	// Make a bunch of concurrent RPCs on the above clientconns. These will
+	// eventually race with Stop(), and will start to fail.
+	var wg sync.WaitGroup
+	for i := 0; i < numChannels; i++ {
+		client := testgrpc.NewTestServiceClient(ccs[i])
+		for j := 0; j < numRPCLoops; j++ {
+			wg.Add(1)
+			go func(client testgrpc.TestServiceClient) {
+				defer wg.Done()
+				for {
+					sCtx, sCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+					defer sCancel()
```

Comment:
This is probably why you're getting the deadline exceeded errors.  Just leave these using the base ctx and it should be fine I think?

Reply:
Oh crap, didn't see this comment. But figure out the same anyways :). Could have saved 30m if I had looked earlier, but I think the investigation helped.